File: js/ai/tests/formats/array_test.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as assert from 'assert';
import { describe, it } from 'node:test';
import { arrayFormatter } from '../../src/formats/array.js';
import { GenerateResponseChunk } from '../../src/generate.js';
import { Message } from '../../src/message.js';
import type {
  GenerateResponseChunkData,
  MessageData,
} from '../../src/model.js';

describe('arrayFormat', () => {
  const streamingTests = [
    {
      desc: 'emits complete array items as they arrive',
      chunks: [
        {
          text: '[{"id": 1,',
          want: [],
        },
        {
          text: '"name": "first"}',
          want: [{ id: 1, name: 'first' }],
        },
        {
          text: ', {"id": 2, "name": "second"}]',
          want: [{ id: 2, name: 'second' }],
        },
      ],
    },
    {
      desc: 'handles single item arrays',
      chunks: [
        {
          text: '[{"id": 1, "name": "single"}]',
          want: [{ id: 1, name: 'single' }],
        },
      ],
    },
    {
      desc: 'handles preamble with code fence',
      chunks: [
        {
          text: 'Here is the array you requested:\n\n```json\n[',
          want: [],
        },
        {
          text: '{"id": 1, "name": "item"}]\n```',
          want: [{ id: 1, name: 'item' }],
        },
      ],
    },
  ];

  for (const st of streamingTests) {
    it(st.desc, () => {
      const parser = arrayFormatter.handler();
      const chunks: GenerateResponseChunkData[] = [];

      for (const chunk of st.chunks) {
        const newChunk: GenerateResponseChunkData = {
          index: 0,
          role: 'model',
          content: [{ text: chunk.text }],
        };

        const result = parser.parseChunk!(
          new GenerateResponseChunk(newChunk, {
            index: 0,
            role: 'model',
            previousChunks: chunks,
          })
        );
        chunks.push(newChunk);

        assert.deepStrictEqual(result, chunk.want);
      }
    });
  }

  const messageTests = [
    {
      desc: 'parses complete array response',
      message: {
        role: 'model',
        content: [{ text: '[{"id": 1, "name": "test"}]' }],
      },
      want: [{ id: 1, name: 'test' }],
    },
    {
      desc: 'parses empty array',
      message: {
        role: 'model',
        content: [{ text: '[]' }],
      },
      want: [],
    },
    {
      desc: 'parses array with preamble and code fence',
      message: {
        role: 'model',
        content: [{ text: 'Here is the array:\n\n```json\n[{"id": 1}]\n```' }],
      },
      want: [{ id: 1 }],
    },
  ];

  for (const rt of messageTests) {
    it(rt.desc, () => {
      const parser = arrayFormatter.handler();
      assert.deepStrictEqual(
        parser.parseMessage(new Message(rt.message as MessageData)),
        rt.want
      );
    });
  }

  const errorTests = [
    {
      desc: 'throws error for non-array schema type',
      request: {
        messages: [],
        output: {
          schema: { type: 'string' },
        },
      },
      wantError: /Must supply an 'array' schema type/,
    },
    {
      desc: 'throws error for object schema type',
      request: {
        messages: [],
        output: {
          schema: { type: 'object' },
        },
      },
      wantError: /Must supply an 'array' schema type/,
    },
  ];

  for (const et of errorTests) {
    it(et.desc, () => {
      assert.throws(() => {
        arrayFormatter.handler(et.request);
      }, et.wantError);
    });
  }
});

File: py/packages/genkit/tests/genkit/blocks/formats/array_test.py
#!/usr/bin/env python3
#
# Copyright 2025 Google LLC
# SPDX-License-Identifier: Apache-2.0

"""Tests for the Array format."""

import pytest

from genkit.blocks.formats.array import ArrayFormat
from genkit.blocks.model import GenerateResponseChunkWrapper, MessageWrapper
from genkit.core.error import GenkitError
from genkit.core.typing import GenerateResponseChunk, Message, Part, TextPart


class TestArrayFormatStreaming:
    """Test streaming chunk parsing."""

    def test_emits_complete_array_items_as_they_arrive(self) -> None:
        """Test that complete objects are emitted as they arrive in chunks."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle({'type': 'array', 'items': {'type': 'object'}})

        # Chunk 1: [{"id": 1,
        chunk1 = GenerateResponseChunk(content=[Part(root=TextPart(text='[{"id": 1,'))])
        result1 = fmt.parse_chunk(GenerateResponseChunkWrapper(chunk1, index=0, previous_chunks=[]))
        assert result1 == []

        # Chunk 2: "name": "first"}
        chunk2 = GenerateResponseChunk(content=[Part(root=TextPart(text='"name": "first"}'))])
        result2 = fmt.parse_chunk(GenerateResponseChunkWrapper(chunk2, index=0, previous_chunks=[chunk1]))
        assert result2 == [{'id': 1, 'name': 'first'}]

        # Chunk 3: , {"id": 2, "name": "second"}]
        chunk3 = GenerateResponseChunk(content=[Part(root=TextPart(text=', {"id": 2, "name": "second"}]'))])
        result3 = fmt.parse_chunk(GenerateResponseChunkWrapper(chunk3, index=0, previous_chunks=[chunk1, chunk2]))
        assert result3 == [{'id': 2, 'name': 'second'}]

    def test_handles_single_item_arrays(self) -> None:
        """Test parsing a single item array in one chunk."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle({'type': 'array', 'items': {'type': 'object'}})

        chunk = GenerateResponseChunk(content=[Part(root=TextPart(text='[{"id": 1, "name": "single"}]'))])
        result = fmt.parse_chunk(GenerateResponseChunkWrapper(chunk, index=0, previous_chunks=[]))
        assert result == [{'id': 1, 'name': 'single'}]

    def test_handles_preamble_with_code_fence(self) -> None:
        """Test parsing array with preamble text and code fence."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle({'type': 'array', 'items': {'type': 'object'}})

        # Chunk 1: preamble with code fence start
        chunk1 = GenerateResponseChunk(
            content=[Part(root=TextPart(text='Here is the array you requested:\n\n```json\n['))]
        )
        result1 = fmt.parse_chunk(GenerateResponseChunkWrapper(chunk1, index=0, previous_chunks=[]))
        assert result1 == []

        # Chunk 2: the actual data
        chunk2 = GenerateResponseChunk(content=[Part(root=TextPart(text='{"id": 1, "name": "item"}]\n```'))])
        result2 = fmt.parse_chunk(GenerateResponseChunkWrapper(chunk2, index=0, previous_chunks=[chunk1]))
        assert result2 == [{'id': 1, 'name': 'item'}]


class TestArrayFormatMessage:
    """Test complete message parsing."""

    def test_parses_complete_array_response(self) -> None:
        """Test parsing a complete array response."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle({'type': 'array', 'items': {'type': 'object'}})

        result = fmt.parse_message(
            MessageWrapper(Message(role='model', content=[Part(root=TextPart(text='[{"id": 1, "name": "test"}]'))]))
        )
        assert result == [{'id': 1, 'name': 'test'}]

    def test_parses_empty_array(self) -> None:
        """Test parsing an empty array."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle({'type': 'array', 'items': {'type': 'object'}})

        result = fmt.parse_message(MessageWrapper(Message(role='model', content=[Part(root=TextPart(text='[]'))])))
        assert result == []

    def test_parses_array_with_preamble_and_code_fence(self) -> None:
        """Test parsing array with preamble and code fence."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle({'type': 'array', 'items': {'type': 'object'}})

        result = fmt.parse_message(
            MessageWrapper(
                Message(
                    role='model', content=[Part(root=TextPart(text='Here is the array:\n\n```json\n[{"id": 1}]\n```'))]
                )
            )
        )
        assert result == [{'id': 1}]


class TestArrayFormatErrors:
    """Test error handling."""

    def test_throws_error_for_non_array_schema_type(self) -> None:
        """Test that non-array schema type raises error."""
        array_fmt = ArrayFormat()

        with pytest.raises(GenkitError) as exc_info:
            array_fmt.handle({'type': 'string'})
        assert "Must supply an 'array' schema type" in str(exc_info.value)

    def test_throws_error_for_object_schema_type(self) -> None:
        """Test that object schema type raises error."""
        array_fmt = ArrayFormat()

        with pytest.raises(GenkitError) as exc_info:
            array_fmt.handle({'type': 'object'})
        assert "Must supply an 'array' schema type" in str(exc_info.value)


class TestArrayFormatInstructions:
    """Test instruction generation."""

    def test_generates_instructions_with_schema(self) -> None:
        """Test that instructions are generated when schema is provided."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle({'type': 'array', 'items': {'type': 'object'}})

        assert fmt.instructions is not None
        assert 'Output should be a JSON array' in fmt.instructions

    def test_no_instructions_without_schema(self) -> None:
        """Test that no instructions are generated without schema."""
        array_fmt = ArrayFormat()
        fmt = array_fmt.handle(None)

        assert fmt.instructions is None
